Content.Server/Nutrition/EntitySystems/SmokingSystem.Vape.cs

```diff
@@ -1,169 +1,201 @@
         {
-            args.Handled = true;
+            _explosionSystem.QueueExplosion(uid, "Default", vape.ExplosionIntensity, 0.5f, 3, canCreateVacuum: false);
+            EntityManager.QueueDeleteEntity(uid);
+            return;
         }
+
+        var doAfterArgs = new DoAfterArgs(EntityManager,
+            user,
+            forced ? vape.Delay : vape.UserDelay,
+            new VapeDoAfterEvent(),
+            uid,
+            target: target,
+            used: uid)
+        {
+            BreakOnMove = forced,
+            BreakOnDamage = true
+        };
+
+        _doAfterSystem.TryStartDoAfter(doAfterArgs);
+    }
+
+    private void OnVapeDoAfter(Entity<VapeComponent> ent, ref VapeDoAfterEvent args)
+    {
+        var (uid, vape) = ent;
+        var user = args.User;
+        var target = args.Args.Target;
+
+        if (args.Cancelled || args.Handled || target == null)
+            return;
+
+        if (!TryComp<BloodstreamComponent>(target, out var bloodstream))
+            return;
+
+        if (!_solutionContainerSystem.TryGetRefillableSolution(uid, out var solutionComp, out var solution))
+            return;
+
+        if (!_solutionContainerSystem.ResolveSolution(target.Value, bloodstream.ChemicalSolutionName, ref bloodstream.ChemicalSolution, out var chemSolution))
+            return;
+
+        var environment = _atmos.GetContainingMixture(target.Value, true, true);
+
+        if (environment == null)
+            return;
+
+        var merger = new GasMixture(1) { Temperature = solution.Temperature };
+        merger.SetMoles(vape.GasType, solution.Volume.Value / vape.ReductionFactor);
+
+        _atmos.Merge(environment, merger);
+
+        // Move as much solution as we can
+        var removedSolution = _solutionContainerSystem.SplitSolution(solutionComp.Value, FixedPoint2.Min(solution.Volume, chemSolution.AvailableVolume));
+
+        // Make vape's solution affect entity
+        _bloodstreamSystem.TryAddToChemicals(target.Value, removedSolution, bloodstream);
+
+        _reactiveSystem.DoEntityReaction(target.Value, removedSolution, ReactionMethod.Ingestion);
+
+        // Clean vape's solution
+        _solutionContainerSystem.RemoveAllSolution(solutionComp.Value);
+
+        // Smoking kills (your lungs, but there is no organ damage yet)
+        _damageableSystem.TryChangeDamage(target.Value, vape.Damage, true);
+
+        if (args.User != args.Target)
+        {
+            var targetName = Identity.Entity(target.Value, EntityManager);
+            var userName = Identity.Entity(user, EntityManager);
+
+            _popupSystem.PopupEntity(
+                Loc.GetString("vape-component-vape-success-forced", ("user", userName)), target.Value,
+                target.Value);
+
+            _popupSystem.PopupEntity(
+                Loc.GetString("vape-component-vape-success-user-forced", ("target", targetName)), user,
+                target.Value);
```

Comment:
```suggestion
                user);
```
you show both popups to the target